### PR TITLE
Reader: Use Gutenberg for Reader Comments via IsolatedBlockEditor

### PR DESCRIPTION
**Pull request by sarayourfriend**
#### Changes proposed in this Pull Request

* Uses the `IsolatedBlockEditor` to enable Gutenberg for Reader comments

Caveats of the current version: Autocompleters do not work. I have no idea why but it appears the filter isn't running as the `console.log` in the `addAutocompleters` function isn't running ever (even though this worked on previous proof of concepts). Ideas I've explored:
1. Maybe it's calling a different `addFilter` than needed (as in package resolution issues) but it's not that. The packages are deduplicated and it's using the same verison of `@wordpress/hooks` package for everything.
2. Maybe the way to do this changed since I last explored it? Nope. It's the same in the private p2editor repository so I know it is the correct way to do it.
3. Maybe the promise isn't resolving for some reason? Nope, I've verified the promise is resolving as expected.

Anyone have clever ideas of how to debug this issue? It seems to be something down to the way the filter is being added because it's just not being run anymore.

I also expect this will blow up the bundle size somehow. There are some significant new dependencies, but that's why I decided to `AsyncLoad` the comments box. This should be fine as there's on most posts you would have to scroll down to see it anyway so it's okay if it pops into view later.

#### Testing instructions

* Open up a blog post in reader, append `flags=reader/gutenberg-for-comments` and try to leave a comment (it works as expected aside from auto completers 😁)
* Also leave a comment without the flag to make sure that the old way still works (with user mentions and paste to link).

Related to #48386

### CODE REVIEW

**Comment by sarayourfriend** [comment]
Part of the build fails (not sure why it only fails in some cases) due to a missing z-index. I think this is because of a breaking change in the deduplicated `@wordpress` scope of packages 😱

**Review by tyxla — COMMENTED** [review]
This is wonderful work @sarayourfriend ❤️  It's great to see this going so far!

I have a couple of thoughts and suggestions here:

* If you're interested in that, it seems like we could split the PR into self-contained pieces, for example:
  * Upgrading any `@wordpress/` packages and adding the isolated block editor as a dependency.
  * Adding the `block-editor` component in `comments` (with the autocompleters) - looks like a good place to debug the issue you described.
* Allowing the current comment editor and the blocked editor one to co-exist. That would mean, enabling the block editor one under a feature flag. That would not only allow us to merge this PR (piece by piece or not), but it would then allow us to further iterate on the block editor discussion experience, while not disturbing anything on production. Later, when things look good, this strategy would allow us to smoothly remove the good old plain comment form in favor of the new shiny one.

What do you think?

**Comment by sarayourfriend** [comment]
Definitely @tyxla, this PR is no where near mergable in its current state.

Once the #50922 PR drops it should smooth some of these things.

Adding a feature flag sounds like a great idea.

**Comment by sarayourfriend** [comment]
@tyxla @jsnajdr can one of y'all take a look at this and help me diagnose the increase in bundle size for Gutenboarding? I don't think Gutenboarding should be affected by the changes, even the additional dependencies.

**Comment by sarayourfriend** [comment]
I've done some digging into the bundle size increase and I'm even more confused now. When I run `yarn analyse-bundles` on this branch and `trunk`, the Gutenboarding entrypoint is actually _smaller_ on this branch (which makes sense, probably some things are being shared that weren't being shared before) but that's in direct contradiction to what ICFY is telling us.

I'm stumped at this point, any help is appreciated ❤️

**Comment by tyxla** [comment]
> I've done some digging into the bundle size increase and I'm even more confused now. When I run `yarn analyse-bundles` on this branch and `trunk`, the Gutenboarding entrypoint is actually _smaller_ on this branch (which makes sense, probably some things are being shared that weren't being shared before) but that's in direct contradiction to what ICFY is telling us.
> 
> I'm stumped at this point, any help is appreciated ❤️

@sarayourfriend I have the same observations when testing locally. To me this seems like a bug with how ICFY is comparing bundles - we've seen it before with other branches under some circumstances. I'll defer to @jsnajdr to take a look, though, because I'm not that familiar with ICFY and how it could potentially misbehave here.

**Comment by sarayourfriend** [comment]
@jsnajdr When you have a chance to take a look at this and give it a review that'd be super helpful ❤️

File: client/blocks/comments/block-editor/autocompleters/user.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+				if ( error ) return reject( error );
+
+				return resolve(
+					Object.entries( result.list ).map( ( [ subdomain, p2 ] ) => ( {
+						...p2,
+						subdomain,
+					} ) )
+				);
+			}
+		);
```
Comment (jsnajdr) [inline client/blocks/comments/block-editor/autocompleters/user.jsx]: Let's use the promise API of `wpcom.req.get` here:
```js
await wpcom.req.get( { path } ).then( result => Object.entries( result.list ).map( ... ) )
```

File: client/blocks/comments/block-editor/index.jsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+			addFilter(
+				'editor.Autocomplete.completers',
+				'readerComments/autocompleters',
+				addAutoCompleters
+			);
+		} );
+	}, [ suggestions ] );
+
+	return (
+		<>
```
Comment (jsnajdr) [inline client/blocks/comments/block-editor/index.jsx]: nit: do we need the fragment here? There is only one child component.

**Review by jsnajdr — APPROVED** [review]
> the Gutenboarding entrypoint is actually smaller on this branch (which makes sense, probably some things are being shared that weren't being shared before) but that's in direct contradiction to what ICFY is telling us.

I didn't find any contradiction between ICFY and a local build. The entrypoint size is not just the size of the `entry-gutenboarding` chunk. It's also 16 other shared chunks that are shared with other chunk groups.

In total, the `entry-gutenboarding` chunk group got approx 50kB bigger, which is approx 10% of its total size. There's one extra 50kB shared chunks, containing mostly code from `@wordpress/dom`, `@wordpress/element` and similar Gutenberg libraries.

I don't have any clear explanation about why exactly this happened. A lot of Gutenberg code is now used in two chunk groups instead of just one, and webpack algorithm decided to reshuffle the shared chunks in a way that inflates `entry-gutenboarding` by 10%.

Also, when a module is used at two places instead of one, it can no longer be optimized by module concatenation. That's an optimization similar to inlining functions -- only it inlines modules. That's another reason why the result can be bigger.

To summarize, I wouldn't be worried about the bundle size increase. This PR doesn't do anything wrong. Another big PR like this can cause a webpack reshuffle that makes `entry-gutenboarding` smaller again. It's essentially a random event and its effects should even out over the long term.

**Comment by sarayourfriend** [comment]
Thanks for the explanation @jsnajdr! That was very helpful.
